K3po.NUnit/src/ScriptRunner.cs

```diff
@@ -134,29 +145,150 @@ public void Start()
+                return null;
         }
 
         public void Abort()
         {
             _abortScheduled = true;
+            SendAbortCommand();
             _latch.NotifyAbort();
         }
 
         private void SendAbortCommand()
         {
             AbortCommand abortCommand = new AbortCommand();
-            _control.WriteCommand(abortCommand);
+            _controller.WriteCommand(abortCommand);
+        }
+
+        public void Start()
+        {
+            if (_latch.IsPrepared)
+            {
+                StartCommand startCommand = new StartCommand();
+                _controller.WriteCommand(startCommand);
+            }
+            else
+            {
+                throw new InvalidOperationException("K3po is not ready start");
+            }
+        }
+
+        public void AwaitBarrier(String barrierName)
+        {
+            if (!_barriers.Keys.Contains(barrierName))
+            {
+                throw new ArgumentException(string.Format(
+                        "Barrier with {0} is not present in the script and thus can't be waited upon", barrierName));
+            }
+            CountdownEvent notifiedEvent = new CountdownEvent(1);
+            BarrierStateMachine barrierStateMachine = _barriers[barrierName];
+            barrierStateMachine.AddListener(new BarrierStateListener1(notifiedEvent));
+
+            try
+            {
+                _controller.Await(barrierName);
+            }
+            catch (Exception e)
+            {
+                _latch.NotifyException(e);
+            }
+            notifiedEvent.Wait();
+        }
+
+        public void NotifyBarrier(string barrierName)
+        {
+            if (!_barriers.ContainsKey(barrierName))
+            {
+                throw new ArgumentException(string.Format("Barrier with {0} is not present in the script and thus can't be notified", barrierName));
+            }
+            CountdownEvent notified = new CountdownEvent(1);
+            BarrierStateMachine barrierStateMachine = _barriers[barrierName];
+            barrierStateMachine.AddListener(new BarrierStateListener2(this, barrierName, notified));
+            notified.Wait();
+        }
+
+        public void Dispose()
+        {
+            _controller.Dispose();
+            try
+            {
+                CommandEvent commandEvent = _controller.ReadEvent();
+
+                // ensure it is the correct event
+                switch (commandEvent.EventKind)
+                {
+                    case CommandEvent.Kind.DISPOSED:
+                        _latch.NotifyDisposed();
+                        break;
+                    default:
+                        throw new ArgumentException("Unrecognized event kind: " + commandEvent.EventKind);
+                }
+            }
+            finally
+            {
+                _controller.Disconnect();
+            }
+        }
+
+        class BarrierStateListener1 : BarrierStateListener
+        {

```

Comment:
Can you rename BarrierStateListener1 and BarrierStateListener2 to be a little more self explanatory on when to use one vs the other (i.e NotifyBarrierStateListener and AwaitBarrierStateListener or something like that)